api/src/file/epoll.rs

```diff
@@ -60,189 +123,306 @@ impl PartialEq for EntryKey {
-            (
-                Some(EpollEvent {
-                    events,
-                    user_data: self.event.user_data,
-                }),
-                !self
-                    .flags
-                    .intersects(EpollFlags::EDGE_TRIGGER | EpollFlags::ONESHOT),
-            )
-        } else {
-            (None, false)
+    #[inline]
+    fn is_in_queue(&self) -> bool {
+        self.in_ready_queue.load(Ordering::Acquire)
+    }
+
+    #[inline]
+    fn try_mark_in_queue(&self) -> bool {
+        self.in_ready_queue
+            .compare_exchange(false, true, Ordering::AcqRel, Ordering::Acquire)
+            .is_ok()
+    }
+
+    #[inline]
+    fn mark_not_in_queue(&self) {
+        self.in_ready_queue.store(false, Ordering::Release);
+    }
+
+    fn consume(&self, file: &dyn FileLike) -> ConsumeResult {
+        let current_events = file.poll();
+        let matched = current_events & self.event.events;
+
+        // not ready
+        if matched.is_empty() {
+            return ConsumeResult::NoEvent;
+        }
+
+        let mut mode = self.mode.lock();
+        let (should_notify, new_mode) = mode.should_notify();
+        *mode = new_mode;
+        trace!(
+            "consume fd: {} matches {:?} should notify: {} ",
+            self.key.fd, matched, should_notify
+        );
+
+        if !should_notify {
+            return ConsumeResult::NoEvent;
+        }
+
+        // create event
+        let event = EpollEvent {
+            events: matched,
+            user_data: self.event.user_data,
+        };
+
+        // shoud still keep in ready?
+        match *mode {
+            TriggerMode::Level => ConsumeResult::EventAndKeep(event),
+            TriggerMode::Edge { .. } | TriggerMode::OneShot { .. } => {
+                ConsumeResult::EventAndRemove(event)
+            }
         }
     }
 }
 
-#[derive(Default)]
-pub struct Epoll {
+struct InterestWaker {
+    epoll: Weak<EpollInner>,
+    interest: Weak<EpollInterest>,
+}
+
+impl Wake for InterestWaker {
+    fn wake(self: Arc<Self>) {
+        self.wake_by_ref();
+    }
+
+    fn wake_by_ref(self: &Arc<Self>) {
+        let Some(epoll) = self.epoll.upgrade() else {
+            return;
+        };
+
+        let Some(interest) = self.interest.upgrade() else {
+            return;
+        };
+
+        if interest.try_mark_in_queue() {
+            epoll
+                .ready_queue
+                .lock()
+                .push_back(Arc::downgrade(&interest));
+            trace!(
+                "Epoll: fd={} added to ready queue, events={:?} wake up poller",
+                interest.key.fd, interest.event.events
+            );
+            epoll.poll_ready.clone().wake();
+        }
+    }
+}
+
+struct EpollInner {
     interests: SpinNoPreempt<HashMap<EntryKey, Arc<EpollInterest>>>,
-    ready: Arc<SpinNoPreempt<ReadyList>>,
+    ready_queue: SpinNoPreempt<VecDeque<Weak<EpollInterest>>>,
     poll_ready: Arc<PollSet>,
```

Comment:
```suggestion
    poll_ready: PollSet,
```

This `Arc` seems to be unnecessary now.

Reply:
Done